keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+### Legacy plugin migration
+
+Although legacy plugins are still in alpha, the alpha has been around in standalone Kustomize for quite a while. Documentation on the migration process will be provided. See the [Rollout Plan](#rollout-plan) section for more detail on deprecation and removal of the existing alphas.
+
+The exec plugin conversion process should be relatively straightforward: copy the provider in-tree and update the configuration that referred to it to have a Provider stanza pointing to the new location. `argsFromFile` and `argsOneLiner` can be straightforwardly converted to the `args` subfield the KRM Functions style already supports.
+
+For authors who build plugins for distribution (as opposed to bespoke plugins where embedding within an end-user Kustomization is an appropriate solution), the recommendation will be to upgrade to the Catalog model. A Catalog guide will be added to the Kustomize website and will address the scenario of migrating a legacy style plugin.
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an end user, I want to develop a simple plugin for my particular Kustomization.
+
+1. Write a script that adheres to the [KRM Functions Specification] and place it within the Kustomization root.
+2. Write a corresponding plugin config and reference it from the Kustomization's generator or transformer field as appropriate.
+3. Run `kustomize edit generate-catalog`
+3. Run `kustomize build --trusted-plugin-catalog=catalog.yaml`
+
+```bash
+# Possible directory structure
+.
+⊢ Kustomization.yaml
+⊢ input.yaml
+⊢ reorder.yaml
+⊢ plugin-catalog.yaml
+∟ plugins
+  ∟ sorter
+    ⊢ sorter.rb
+    ∟ sorter_test.rb
+```
+
+```yaml
+# Example Kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- input.yaml
+transformers:
+- reorder.yaml
+```
+
+```yaml
+# Example reorder.yaml
+apiVersion: local-config.example.co/v1
+kind: CustomSorter
+provider:
+  exec: 
+  	path: plugins/sorter/sorter.rb
+spec:
+  order: NsGKVN
+```
+
+```yaml
+# Example catalog.yaml (auto-generated)
+
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata:
+  name: "local-plugins"
+spec:
+  modules:
+  - apiVersion: local-config.example.co/v1
+    kind: CustomSorter
+    provider:
+      exec:
+        - path: plugins/sorter/sorter.rb
+          sha256: [a hash]
+```
+
+
+#### Story 2
+
+As an end user, I want to use one or more plugins published by a third party I trust.
+
+1. Write the config objects for the plugins you want to use, and reference them from your Kustomization's generators or transformers field as appropriate.
+2. Run `kustomize build --trusted-plugin-catalog=https://catalog.kpt.dev/1.2.3.json`
```

Comment:
As I commented in https://github.com/kubernetes/enhancements/pull/2908/files#r704953434, IMO there are user stories missing.

I imagine we can collaboratively build a sig-cli public KRM function catalog (repo may be in kubernetes-sigs) and a website that may be similar to https://catalog.kpt.dev/.
Each function have a [KRMFunction.yaml(or json)](https://github.com/kubernetes/enhancements/pull/2908/files#r702092083). e.g. https://catalog.kpt.dev/starlark/v0.2/KRMFunction.yaml.
If needed, the catalog site can serve a catalog.yaml (or json). e.g. https://catalog.kpt.dev/catalog.yaml.

As a platform developer at enterprise company Example Co, I can create a catalog containing the plugins built by others (e.g. vendors I trust). I can probably create a catalog CR like:
```yaml
apiVersion: config.kubernetes.io/v1alpha1
kind: Catalog
metadata: 
  ...
spec: 
  modules: 
    - providerRef: https://catalog.kpt.dev/starlark/v0.2/KRMFunction.yaml
       ...
    - providerRef: https://catalog.kpt.dev/gatekeeper/v0.2/KRMFunction.yaml
       ...
    - providerRef: https://catalog.kpt.dev/kubeval/v0.1/KRMFunction.yaml
       ...
```

Reply:
Since that's about the details of what a Catalog looks like, let's have that discussion on the other KEP. But FWIW at a high level I think it's a good idea for Catalog to be able to support plugins from heterogeneous sources in some way.